rln-research/membership-merkle-tree-storage-evaluation.md

```diff
@@ -0,0 +1,107 @@
+Note that this analysis is only valid for the storage overhead, however, setting a higher value for the depth may have a different impact on the computational load. This needs to be evaluated separately.
+
+
+
+### FMT vs OMT
+
+Comparing the max storage overhead of FMT and OMT, as depicted in the figure below, brings us to the conclusion that FMT is more efficient than OMT. However, this is only true for the worst-case scenario where the group is full to the maximum tree capacity.
+![Max relative](https://i.imgur.com/G8COahY.png)
+
+But, this worst-case scenario does not always hold, instead, in most states of the system, the group size is much less than the tree capacity. In the following figure, we take a snapshot of the system at group size `1000`. In this snapshot, we compare the storage overhead incurred by FMT and OMT for `d=[10..20]`.
+![group 1000](https://i.imgur.com/NuIDALV.png)
+
+As you can see, in this more realistic scenario, FMT shows better performance than OMT. For example, for `d=20`, persisting a group of `1000` members using OMT requires  `0.097 MB` whereas it requires `67.1 MB` when using FMT, `700` times more than OMT.
+
+It brings us to another conclusion that OMT outperforms FMT when group size is less than the tree capacity. However, the exact point at which this relationship changed needs more investigation.
+
+
+
+
+
+## Conclusion
+1) Storage overhead of membership MT does not seem to be an issue for desktop nodes given that the maximum overhead to persist an MT for a large group of `2^20` and `2^10` members does not exceed `100 MB` (see below for more accurate numbers).
+
+
+| Algorithm\Tree depth | d=10      | d=20      |
+| -------------------- | --------- | --------- |
+| FMT                  | 33.55  MB | 67.10 MB  |
+| OMT                  | 50.33  MB | 100.66 MB |
+
+However, the MT storage overhead may not be reasonable for mobile devices especially if we plan to load the MT in memory in which case we will/may face the per-app memory cap of `32 MB` on Android phones.
+
+2) If we can not predict the maximum group size (hence the max tree depth), then it is more efficient to use the OMT algorithm. This is because the storage overhead of OMT is proportional to the group size and less affected by the tree depth.  For instance, consider a tree with a depth of `20` and with `1000` active leaves. MT storage using OMT requires  `0.097 MB` whereas  `67.1 MB` when using FMT, `700` times more than OMT.  However, the exact point at which this relationship may change needs more investigation.
```

Comment:
This is indeed an interesting point

I think using OMT make sense, though we should be aware of the possibility of OOM and have some fallback plan for scalable production usage. Maybe there are some disk-based tweaks we can use here.